MC-29776 Added PHPStan information (#6329)
magento/devdocs: src/guides/v2.3/test/testing.md

> Please see the [`magento dev:tests:run`][] documentation for more information, using the test type `static`.

Reviewer: ```suggestion
See the [`magento dev:tests:run`][] documentation for more information, using the test type `static`.
```

Author: The shame.

### PHPStan

> [PHPStan][] is a static analysis tool for PHP. Magento now integrates PHPStan by default.

Reviewer: ```suggestion
[PHPStan][] is a static analysis tool for PHP. Magento integrates PHPStan by default.
```

I know we use words like 'now' in release notes, but seems it could be weird in regular documentation. In a year, now is not the intended now? Just a thought.

-  Anonymous function has an unused use
-  Array has N duplicate keys
-  Call to an undefined method

#### Suppressing errors

To ignore errors in a particular piece of code, use the `phpstan:ignore` annotation.

```php
// phpstan:ignore "Error message"
```

The error message text is optional.
It is possible to use the `*` wildcard within messages.

```php
/**
 * Class ClassWithIgnoreAnnotation
 */
class ClassWithIgnoreAnnotation
{
    /**
     * Test method.
     * phpstan:ignore "Method level error"
     */
    public function getProductList()
    {
        // phpstan:ignore "Method Magento\PhpStan\Formatters\Fixtures\ClassWithIgnoreAnnotation::testMethod() invoked with 2 parameters, 1 required."
        $this->testMethod('test1', 'test2');
        // phpstan:ignore "Method * invoked with 2 parameters, 1 required."
        $this->testMethod('test1', 'test2');
        // phpstan:ignore
        $this->testMethod('test1', 'test2');
        $this->testMethod('test1', 'test2'); // phpstan:ignore
    }
    /**
     * @param string $arg1
     * @return string
     */
    private function testMethod(string $arg1)
    {
        return $arg1;
    }
}
```

> #### Adding files/directories to PHPStan blacklist

Reviewer: why use a forward slash in a title?

Author: Good point.

If some files are broken on purpose, add them to the blacklist to ignore them.

The blacklist is located at `dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/blacklist/`.

```config
lib/internal/Magento/Framework/Interception/Test/Unit/Config/ConfigTest.php
lib/internal/Magento/Framework/Cache/Backend/Eaccelerator.php
```

#### Exclude error messages from the analysis results

If some error messages are not relevant for the current testing scenario, use regular expressions to filter them.

The PHPStan configuration file is `dev/tests/static/testsuite/Magento/Test/Php/_files/phpstan/phpstan.neon`.

Example:

```config
ignoreErrors:
    # Ignore PHPStan\Rules\Classes\UnusedConstructorParametersRule
   - '#Constructor of class [a-zA-Z0-9\\_]+ has an unused parameter#'
```

> #### Exclude directories from analysis using the `excludes_analyse` option

Reviewer: This says directories, but in the example it could be files, too.

```suggestion
#### Exclude files and directories from analysis
```

> As needed in some cases, directories and files can be excluded via the PHPStan configuration file:

Reviewer: Via is on a poo list somewhere.
https://wiki.corp.magento.com/display/MDOC/Style+guide+-+grammar#Styleguide-grammar-Wording

```suggestion
As needed in some cases, you can exclude specific directories and files using the `excludes_analyse` section in the PHPStan configuration file:
```
